src/fetch-groups.js

```diff
@@ -0,0 +1,157 @@
+/**
+ * Module that exports a function that takes a list of specifications as input
+ * and computes, for each of them, the name of the organization and groups
+ * within that organization that develop the specification.
+ *
+ * The function needs an authentication token for the GitHub API as well as for
+ * the W3C API.
+ */
+
+const fetch = require("node-fetch");
+const { Octokit } = require("@octokit/rest");
+
+
+/**
+ * Exports main function that takes a list of specs (with a url property)
+ * as input, completes entries with an "organization" property that contains the
+ * name of the organization such as W3C, WHATWG, IETF, Khronos Group,
+ * Ecma International, and a "groups" property that contains an array of objects
+ * that describe the groups responsible for the spec.
+ * 
+ * The function preserves the properties if they have already been provided in
+ * the input array.
+ *
+ * The options parameter is used to specify the GitHub API and W3C API
+ * authentication tokens.
+ */
+module.exports = async function (specs, options) {
+  // Maintain a cache of fetched resources in memory to avoid sending the
+  // same fetch request again and again
+  const cache = {};
+
+  // Helper function to retrieve a JSON resource or return null if resource
+  // cannot be retrieved
+  async function fetchJSON(url, options) {
+    const body = cache[url] ?? await fetch(url, options).then(res => {
+      if (res.status !== 200) {
+        throw new Error(`W3C API returned an error, status code is ${res.status}`);
+      }
+      return res.json();
+    });
+    cache[url] = body;
+    return body;
+  }
+
+  const w3cOptions = options?.w3cApiKey ? { headers: {
+    Authorization: `W3C-API apikey="${options.w3cApiKey}"`
+  }} : {};
+
+  for (const spec of specs) {
+    // Note: same RegExp logic to detect WHATWG, TC39, SVG, Khronos Group and
+    // GitHub URLs as in compute-repository.
+    const whatwg = spec.url.match(/^https:\/\/([^\.]*).spec.whatwg.org\//);
+    if (whatwg) {
+      spec.organization = spec.organization ?? "WHATWG";
+      spec.groups = spec.groups ?? [{
+        name: "WHATWG",
```

Comment:
the WHATWG has a notion of [workstreams](https://whatwg.org/workstream-policy#workstream) - workstreams have “participants”, so maybe they're the closest thing to a notion of group in that context.

Reply:
The [list of workstreams](https://whatwg.org/workstreams) is generated from a [WHATWG-SG maintained JSON file](https://github.com/whatwg/sg/blob/main/db.json)

Reply:
So, it's easy to retrieve the name of the workstream but there are no dedicated URL per workstream as far as I can tell. If we want the `url` property to act as a sort of group identifier, the best is probably to link to the spec itself for now (today, there's a 1-to-1 mapping between workstreams and specs). Any other idea?

Reply:
no other idea, and +1 to using the spec itself for now given the congruence